RFC: Endpoint structure

thockin opened the pull request:
@bgrant0607 @quinton-hoole as discussed this morning.  This will still be a bit awkward to use but no worse than the other forms.  It devolves into a list of subsets each holding 1 IP and the ports on that IP, but in most cases it will be a single subset or maybe 2.

Just looking for nodding that this is what we talked about and maybe feedback on names :)

thockin committed:
RFC: Endpoint structure

# *** ERROR: *** docs are out of sync between cli and markdown
# run hack/run-gendocs.sh > docs/kubectl.md to regenerate

#
# Your commit will be aborted unless you regenerate docs.
    COMMIT_BLOCKED_ON_GENDOCS

quinton-hoole commented on pkg/api/types.go:
```diff
@@ -785,30 +785,53 @@ type Service struct {
+type EndpointAddress struct {
+	// The IP of this endpoint.
 	// TODO: This should allow hostname or IP, see #4447.
 	IP string `json:"ip"`
```
I'd very much like to get away from using strings in places like this, but you know that already.

thockin replied:
There's an open issue to use net.IP or util.IP, but #4447 also wants hostnames.  We need to revisit this en masse.

quinton-hoole commented on pkg/api/types.go:
```diff
@@ -785,30 +785,53 @@ type Service struct {
+	Name string
+
+	// The IP protocol for this port.
+	Protocol Protocol
```
nit: Is there any reason why we use strings under the hood here, rather than enumeration like http://www.iana.org/assignments/protocol-numbers/protocol-numbers.xhtml
?

thockin replied:
Mostly because humans are typing in JSON and YAML code and it is more obvious that these "enums" be string values.

quinton-hoole replied:
Fair enough.  I don't fully grok all of the contact points between the JSON and Go code yet, but do we have a good mechanism for avoiding bugs like the following in the Go code?

endpointPort.Protocol = "TCP"
....
if endpointPort.Protocol != "tcp" {
     // do non-TCP stuff
}

.... other than having code reviewers shouting at people and telling them to use ProtocolTCP instead of "TCP" and "tcp"?  I'd rather have their compiler shout at them :-)

bgrant0607 replied:
Sadly, I don't think we do.

quinton-hoole commented on pkg/api/types.go:
```diff
@@ -785,30 +785,53 @@ type Service struct {
+// EndpointPort is a tuple that describes a single port.
+type EndpointPort struct {
+	// The name of this port (corresponds to ServicePort.Name).
+	Name string
```
Can we use a type here, rather than a bland string again?  What restrictions are there on the name, is it namespaced, etc.

thockin replied:
I think there's an open github issue to make these things more self-documenting.  I'll document it in comments for now, but we can/should revisit the idea, maybe after 1.0

quinton-hoole commented:
Other than the type-safety related nits, LGTM.  
You might consider abbreviating "Address" to "Addr", as things like EndpointAddress start to become borderline unwieldy in a Java-ish sort of way.  But I don't feel very strongly.

quinton-hoole commented:
Oh, and do we want to place any restrictions on name->port bindings to keep them at least not outrageously inconsistent with http://www.iana.org/assignments/service-names-port-numbers/service-names-port-numbers.xhtml

Or are we happy is people do slightly crazy things like:

e := EndpointPort { Name: "ssh", Protocol: "TCP", port:80 }

bgrant0607 commented:
LGTM. I like it.

As a consumer, in the simple use case I could assert that there's exactly one subset. In more complex cases, at least the topology is a lot more obvious.

I tried to think of something better than "subset", but I think that's nice and succinct, and could be easily elaborated by decorating with adjectives like "homogeneous". 

I might put ports above addresses in EndpointSubset, since the port list will be shorter than the address list, in general.

thockin commented:
We generally eschew such abbreviations in the API, but I'll do it if @bgrant0607  is OK with it.

as for crazy names, I don't think we're HAPPY about it, but I don't think it is our job to prevent it.

bgrant0607 commented:
We don't abbreviate in the API.

bgrant0607 commented:
Note that users won't have to type "EndpointAddress" -- that's the Go type, not the json field name.

thockin committed:
API changes, conversions, defaults

# *** ERROR: *** docs are out of sync between cli and markdown
# run hack/run-gendocs.sh > docs/kubectl.md to regenerate

#
# Your commit will be aborted unless you regenerate docs.
    COMMIT_BLOCKED_ON_GENDOCS

thockin committed:
Do pkg/proxy/*

# *** ERROR: *** docs are out of sync between cli and markdown
# run hack/run-gendocs.sh > docs/kubectl.md to regenerate

#
# Your commit will be aborted unless you regenerate docs.
    COMMIT_BLOCKED_ON_GENDOCS

thockin committed:
Do the endpoints controller

# *** ERROR: *** docs are out of sync between cli and markdown
# run hack/run-gendocs.sh > docs/kubectl.md to regenerate

#
# Your commit will be aborted unless you regenerate docs.
    COMMIT_BLOCKED_ON_GENDOCS

thockin committed:
Reset hash state between runs.

bgrant0607 commented:
@thockin Is this PR still needed?

thockin commented:
you can close it.

On Tue, Mar 24, 2015 at 11:24 AM, Brian Grant <notifications@github.com>
wrote:

> @thockin <https://github.com/thockin> Is this PR still needed?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/5731#issuecomment-85631089>
> .
>

gmarek commented:
Can someone elaborate on what happened with this PR?

thockin commented:
The larger pr to implement it fully is coming soon.
On Mar 25, 2015 2:02 AM, "gmarek" <notifications@github.com> wrote:

> Can someone elaborate on what happened with this PR?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/5731#issuecomment-85938819>
> .
>
